netket/graph/lattice.py

```diff
@@ -104,323 +107,349 @@ def get_edges(atoms_positions, cutoff, distance_atol=cutoff_tol):
     return pairs
 
 
-def create_points(basis_vectors, extent, atom_coords, pbc):
+@dataclass
+class LatticeSite:
+    id: int
+    """Integer ID of this site"""
+    position: Array
+    """Real-space position of this site"""
+    cell_coord: Array
+    """Cell coordinates of this site"""
+    inside: bool
+    """TODO: When exactly is this needed?"""
+
+    def __repr__(self):
+        s = ", ".join(map(str, (self.id, self.cell_coord)))
+        return f"LatticeSite({s})"
+
+
+def create_sites(primitives, extent, apositions, pbc):
     shell_vec = _np.zeros(extent.size, dtype=int)
     shift_vec = _np.zeros(extent.size, dtype=int)
     # note: by modifying these, the number of shells can be tuned.
     shell_vec[pbc] = 2
     shift_vec[pbc] = 1
     ranges = tuple([list(range(ex)) for ex in extent + shell_vec])
-    atoms = []
-    cellANDlabel_to_site = {}
+    sites = []
+    cell_coord_to_site = {}
     for s_cell in itertools.product(*ranges):
         s_coord_cell = _np.asarray(s_cell) - shift_vec
         if _np.any(s_coord_cell < 0) or _np.any(s_coord_cell > (extent - 1)):
             inside = False
         else:
             inside = True
-        atom_count = len(atoms)
-        for i, atom_coord in enumerate(atom_coords):
-            s_coord_atom = s_coord_cell + atom_coord
-            r_coord_atom = _np.matmul(basis_vectors.T, s_coord_atom)
-            atoms.append(
-                {
-                    "Label": i,
-                    "cell": s_coord_cell,
-                    "r_coord": r_coord_atom,
-                    "inside": inside,
-                }
+        atom_count = len(sites)
+        for i, atom_coord in enumerate(apositions):
+            s_coord_site = s_coord_cell + atom_coord
+            r_coord_site = _np.matmul(primitives.T, s_coord_site)
+            cell_coord_site = _np.array((*s_coord_cell, i), dtype=int)
+            sites.append(
+                LatticeSite(
+                    id=None,  # to be set later, after sorting all sites
```

Comment:
E.g. here you could define a counter that increases continuously through all the for loops and use that as `id`, which outputs sites in a logical unit-cell-major, basis-minor order

Reply:
I feel since the edges are sorted later anyway, it's okay to just set the `id` there. I'd maybe change the sorting to use cell coords explicitly at that point, as mentioned above.

Reply:
Since you generate the sites in precisely this order, you don't even need sorting!